src/module.c

```diff
@@ -2987,21 +2991,40 @@ int RM_ZsetRangePrev(RedisModuleKey *key) {
  *
  * Return value:
  *
- * The number of fields updated (that may be less than the number of fields
- * specified because of the XX or NX options).
- *
- * In the following case the return value is always zero:
- *
- * * The key was not open for writing.
- * * The key was associated with a non Hash value.
+ * The number of fields updated or deleted. If the
+ * REDISMODULE_HASH_COUNT_INSERTS is set, inserted fields are also counted. The
+ * returned number may be less than the number of fields specified because of
+ * the XX or NX options.
+ *
+ * If the number returned is less than expected (such as zero), since Redis 6.2
+ * `errno` is set as follows:
```

Comment:
i think the condition of `less than expected` is becoming complicated due to the XX and NX flags.
i didn't read the original ticket recently with all the feedback, so i might be wrong..
but i have a feeling that instead of specifying this, we can say that `errno is set when the return value is 0` and that in case that 0 return value was not due to an error but rather due to the XX/NX filtering, errno will be set to ENOENT.

so the user needs to look at errno only if the return value is 0, but in some cases he can discover that it's not an error. (reminds me of EAGAIN in many system calls)

Reply:
Yes "less than expected" is very vague. :) It actually means "less than the number of specified fields" here.

Fine, I'll set errno only if 0 is returned, which is more strait forward, but I'd like to set it to ENOENT for XX and to EEXIST for NX. ENOENT doesn't make sense for NX, since in this case it means that setting a field was prevented because the field already exists. EEXIST is used in this way by commands which don't overwrite files (e.g. `fopen()`, `open()`, etc.).

Reply:
Recalling that the main issue is that 0 is returned when all fields were inserted, no flags given.

I can't come up with reasonable logic for when to set EEXIST without counting would-inserts and would-updates for various combinations of flags, which would clutter the code and make it hard to explain to users, so I'll go with ENOENT in all these cases as you suggested. I think this is fine. See the fixup commit.